pkg/image/registry/imagestreammapping/rest.go

```diff
@@ -90,18 +102,52 @@ func (s *REST) Create(ctx kapi.Context, obj runtime.Object) (runtime.Object, err
 		Image:                image.Name,
 	}
 
-	if !api.AddTagEventToImageStream(stream, tag, next) {
-		// nothing actually changed
-		return &unversioned.Status{Status: unversioned.StatusSuccess}, nil
-	}
+	var result *unversioned.Status
+	var updateConflictErr error
+	retryErr := client.RetryOnConflict(wait.Backoff{Steps: maxRetriesOnConflict}, func() error {
+		lastEvent := api.LatestTaggedImage(stream, tag)
+		if !api.AddTagEventToImageStream(stream, tag, next) {
+			// nothing actually changed
+			result = &unversioned.Status{Status: unversioned.StatusSuccess}
+			return nil
+		}
+		api.UpdateTrackingTags(stream, tag, next)
 
-	api.UpdateTrackingTags(stream, tag, next)
+		_, err := s.imageStreamRegistry.UpdateImageStreamStatus(ctx, stream)
+		if err != nil {
+			if !errors.IsConflict(err) {
+				return err
+			}
+			// If the update conflicts, get the latest stream and check for tag
```

Comment:
> If you get any error, clear stream (so the next iteration of the loop reads it again).

This code relies on not resetting the stream unless it finds there isn't a real conflict.  That logic ensures that it never re-attempts this work without an imagestream with a resourceversion that is known to be consistent with the first one we grabbed.